File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys

sys.setrecursionlimit(10**9)

input = sys.stdin.readline

V = int(input())
graph = [[] for _ in range(V + 1)]
for _ in range(V - 1):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)


visited = [False] * (V + 1)

parent_child = []


def dfs(start):
    visited[start] = True
    for i in graph[start]:
        if not visited[i]:
            parent_child.append((i, start))
            dfs(i)


dfs(1)
parent_child.sort()
for i in parent_child:
    print(i[1])

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.js
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = require('fs').readFileSync(filePath).toString().trim().split('\n');

const N = +input[0];

const graph = [...Array(N + 1)].map(() => []);
const answer = [...Array(N + 1)].map(() => 0);

for (let i = 1; i < N; i++) {
  const [start, end] = input[i].split(' ').map(Number);
  graph[start].push(end);
  graph[end].push(start);
}

function bfs(start) {
  const q = [start];
  answer[start] = 1;
  while (q.length) {
    const parent = q.pop();
    for (const num of graph[parent]) {
      if (answer[num] === 0) {
        q.push(num);
        answer[num] = parent;
      }
    }
  }
}

bfs(1);

console.log(answer.slice(2).join('\n'));
